lxd/device/gpu.go

```diff
@@ -163,3 +185,464 @@ func gpuSelected(device config.Device, gpu api.ResourcesGPUCard) bool {
+		evaluatedHostPath, err := filepath.EvalSymlinks(mount.HostPath)
+		if err != nil {
+			return nil, err
+		}
+
+		if evaluatedHostPath != mount.HostPath && mount.ContainerPath == strings.TrimPrefix(mount.HostPath, rootPath) {
+			indirectSymlinks = append(indirectSymlinks, cdi.SymlinkEntry{Target: strings.TrimPrefix(evaluatedHostPath, rootPath), Link: mount.ContainerPath})
+			mount.ContainerPath = strings.TrimPrefix(evaluatedHostPath, rootPath)
+		}
+
+		nativeDevices[deviceName] = map[string]string{
+			"type":              "disk",
+			"source":            evaluatedHostPath,
+			"path":              mount.ContainerPath,
+			"shift":             shift,
+			"raw.mount.options": chosenOptsStr,
+		}
+	}
+
+	// If the user desires to run a nested docker container inside a LXD container,
+	// the Tegra CSV files also need to be mounted so that the nvidia docker runtime
+	// can be auto-enabled as 'csv' mode.
+	if cdiID.Vendor() == cdi.Nvidia && cdiID.DeviceType() == cdi.IGPU {
+		tegraCSVFilesCandidates := cdi.DefaultNvidiaTegraCSVFiles()
+		tegraCSVFiles := make([]string, 0)
+		for _, candidate := range tegraCSVFilesCandidates {
+			_, err = os.Stat(filepath.Join(rootPath, candidate))
+			if err == nil {
+				tegraCSVFiles = append(tegraCSVFiles, filepath.Join(rootPath, candidate))
+			} else if os.IsNotExist(err) {
+				continue
+			} else {
+				return nil, err
+			}
+		}
+
+		if len(tegraCSVFiles) == 0 {
+			return nil, fmt.Errorf("No CSV files detected for Tegra iGPU")
+		}
+
+		for _, tegraFile := range tegraCSVFiles {
+			splittedDevName = strings.Split(tegraFile, "/")
+			deviceName = splittedDevName[len(splittedDevName)-1]
+			nativeDevices[deviceName] = map[string]string{
+				"type":     "disk",
+				"source":   tegraFile,
+				"path":     strings.TrimPrefix(tegraFile, rootPath),
+				"shift":    shift,
+				"readonly": "true",
+			}
+		}
+	}
+
+	return indirectSymlinks, nil
+}
+
+// constructCDIHook will create a hook entry in a Hooks struct based on the CDI hook.
+func constructCDIHook(hook *specs.Hook, hooks *cdi.Hooks) error {
+	if hook == nil {
+		return fmt.Errorf("CDI hook is nil")
+	}
+
+	if len(hook.Args) < 5 {
+		return fmt.Errorf("CDI hook %q has not enough arguments", fmt.Sprintf("%s %v", hook.HookName, hook.Args))
+	}
+
+	rootPath := ""
+	if shared.InSnap() {
+		rootPath = "/var/lib/snapd/hostfs"
+	}
+
+	switch hook.Args[2] {
+	case "create-symlinks":
+		for i := 4; i < len(hook.Args); i += 2 {
+			entry := strings.Split(hook.Args[i], "::")
+			if len(entry) != 2 {
+				return fmt.Errorf("invalid symlink entry %q", hook.Args[i])
+			}
+
+			// `Link` is always an absolute path and `Target` (a `Link` points to a `Target`) is relative
+			// to the `Link` location in the CDI spec. A resolving operation will be needed to have the absolute
+			// path of the `Target`
+			hooks.Symlinks = append(hooks.Symlinks, cdi.SymlinkEntry{Link: strings.TrimPrefix(entry[0], rootPath), Target: strings.TrimPrefix(entry[1], rootPath)})
```

Comment:
One note about this: This would then depend on our somewhat unique syntax for defining symlinks. I have a todo to revisit this on our end, so if you have any thoughts on what is easier to consume and read, I'd be happy to hear suggestions.